File: 2018/Thang09-2018/24-09-2018/game.pas
Const   ginp='game.inp';
        gout='game.out';
        maxn=100010;

Var     n:longint;
        res:qword;
        a,b:array[0..maxn] of int64;

Procedure Sort(l,r:longint; var d:array of int64);
Var     i,j:longint;
        t,tmp:int64;
begin
        if l>=r then exit;
        i:=l; j:=r;
        t:=d[random(r-l+1)+l];
        repeat
                while d[i]<t do inc(i);
                while d[j]>t do dec(j);
                if i<=j then
                begin
                        tmp:=d[i]; d[i]:=d[j]; d[j]:=tmp;
                        inc(i); dec(j);
                end;
        until i>j;
        sort(l,j,d); sort(i,r,d);
end;

Function Abs(x:int64):int64;
Begin
        if x<0 then abs:=-x else abs:=x;
End;

Function Min(x,y:qword):qword;
Begin
        if x<y then min:=x else min:=y;
End;

Procedure Answer(x,y:int64);
Begin
        if ((x<0) and (y>0)) or ((x>0) and (y<0)) then res:=min(res,qword(abs(x+y)))
        else res:=min(res,qword(abs(x))+qword(abs(y)));
End;

Procedure Enter;
Var     i:longint;
Begin
        readln(n);
        for i:=1 to n do read(a[i]);
        sort(1,n,a); readln;
        for i:=1 to n do read(b[i]);
        sort(1,n,b);
End;

Procedure Process;
Var     i,j:longint;
Begin
        res:=high(qword); j:=n;
        for i:=1 to n do
        begin
                while (j>0) and (b[j]>=-a[i]) do dec(j);
                if j>0 then answer(a[i],b[j]);
                if j<n then answer(a[i],b[j+1]);
        end;
        write(res);
End;

Begin
        Assign(input,ginp); Assign(output,gout);
        Reset(input); Rewrite(output);
        Enter;
        Process;
        Close(input); Close(output);
End.

File: 2019/Thang03_2019/25_03_2019/game.cpp
#include <bits/stdc++.h>
#define forn(i,a,b) for(int i=(a); i<=(b); i++)
#define repn(i,a,b) for(int i=(a); i<(b); i++)
using namespace std;
//////

const
    int maxn=1100;
int n, m, B;
int a[maxn][maxn], f[maxn][maxn];
int res;
//////

void enter()
{
    scanf("%d%d\n%d\n", &n, &m, &B);
    forn(i, 1, n)
    {
        forn(j, 1, m) scanf("%d", &a[i][j]);
        scanf("\n");
    }
}
//////

void process()
{
    forn(j, 0, m) f[0][j]=B;
    forn(i, 1, n)
    {
        f[i-1][0]=maxn*1e5; f[i-1][m+1]=maxn*1e5;
        forn(j, 1, m)
        {
            f[i][j]=min(f[i-1][j], min(f[i-1][j-1], f[i-1][j+1]))+a[i][j];
            if (a[i][j]==0) f[i][j]-=f[i][j]/2;
        }
    }
    res=maxn*1e5;
    forn(j, 1, m) res=min(res, f[n][j]);
    cout << res<< '\n';

    forn(j, 0, m) f[0][j]=B;
    forn(i, 1, n)
    {
        f[i-1][0]=0; f[i-1][m+1]=0;
        forn(j, 1, m)
        {
            f[i][j]=max(f[i-1][j], max(f[i-1][j-1], f[i-1][j+1]))+a[i][j];
            if (a[i][j]==0) f[i][j]-=f[i][j]/2;
        }
    }
    res=0;
    forn(j, 1, m) res=max(res, f[n][j]);
    cout << res;
}

int main()
{
    freopen("game.inp", "r", stdin);
    freopen("game.out", "w", stdout);
    enter();
    process();
}
